run: Add dry run mode

kyleam opened the pull request:
This PR implements the dry-run operation proposed in gh-5538 by adding a `--dry-run` flag to `run` along with a custom result renderer.  I went light on the reported details, but, if desired, I think it should be straightforward to extend this approach later with more information.

I decided to go with the name `--dry-run` because I think that's the most familiar/obvious name, though `--report` would align with `rerun`.  Also, `--report` might be a bit more natural if this option is later extended to take an optional value (e.g., to restrict or extend what is reported).

---

```
$ touch foo
$ datalad save

$ datalad run --dry-run -i 'fo*' -o out 'cat {inputs[0]} >{outputs[0]}'
Dry run information
 location: /tmp/dl-vaZ53PN
 expanded inputs:
  ['foo']
 expanded outputs:
  ['out']
 command:
  cat {inputs[0]} >{outputs[0]}
 expanded command:
  cat foo >out
```

yarikoptic commented:
Thank you @kyleam - yet to review changes etc, I see myself using it so good.

yarikoptic commented on datalad/core/local/run.py:
```diff
@@ -231,6 +237,13 @@ class Run(Interface):
+            args=("--dry-run",),
+            action="store_true",
+            doc="""Do not run the command; just display details about the
+            command execution."""),
```
I wonder if it is worth adding a note that `glob`s in `input/output`s might not expand (at all or fully) if pointing to some paths within not yet installed subdatasets. 
I do not think we should be getting all those inputs etc,  even though a "perfect" `--dry-run` could have included listing of  (sub)datasets to install, and a number/size of files to get  - I think it would bring up its run time considerably, thus possibly making it less convenient for typical use cases, and thus not done until need is expressed.

mih replied:
Good point, but I fear that the overhead to be able to state this clearly would be substantial. I would lean towards keeping it simple, for a start.

kyleam replied:
I've added a sentence to the docstring mentioning that globs from uninstalled datasets aren't expanded.

yarikoptic commented on datalad/core/local/tests/test_run.py:
```diff
@@ -571,3 +571,40 @@ def test_run_remove_keeps_leading_directory(path):
+
+
+@with_tree(tree={"foo": "f", "bar": "b"})
+def test_dry_run(path):
```
I wonder if worthwhile adding into a test some uninstalled subdataset with run having input and/or output path(s) pointing to it just to ensure that `--dry-run` (or some logic above) doesn't blow somehow due to absent (since subdataset is uninstalled) files, and that the subdataset does not get installed (according to the code it shouldn't but that would be the point of the test to ensure that it will not ;-))

kyleam replied:
Indeed.  Thanks for bring up subdatasets.  I overlooked them when working on this.

yarikoptic reviewed (commented):
Looks good to me.  Didn't try, but left some comments to consider.

mih commented:
First an impulse comment after having played with it for a bit....YES!

This alone will have enormous impact:

```
# wtf?!
% datalad run --dry-run "for m in 1 2 3; do echo $m; done"
Dry run information
 location: /tmp/dummy
 command:
  for m in 1 2 3; do echo ; done
```

```
# ah!
% datalad run --dry-run "for mike in 1 2 3; do echo \$mike > \$mike; done"
Dry run information
 location: /tmp/dummy
 command:
  for m in 1 2 3; do echo $m > $m; done
```

mih commented on datalad/core/local/run.py:
```diff
@@ -245,16 +258,54 @@ def __call__(
+            lines = [fmt_line("location", dry_run_info["pwd_full"])]
+
+            # TODO: Inputs and outputs could be pretty long. These may be worth
+            # truncating.
```
I think the expansion of the inputs/outputs in conjunction with the command is instrumental, not just the command. I gave an example in https://github.com/datalad/datalad/pull/5539#issuecomment-811698301

mih reviewed (commented):
This is great! Thx @kyleam 

I am still struggling with the API and (limited) potential for extending with different reporting strategies, as you already pointed out. I tried a few alternatives that all came out inferior to your's. It would be great to already have this implemented as an option that could take values, but doesn't have to. I have no good approach to suggest right now, though...

kyleam commented:
Thanks for the feedback, @yarikoptic and @mih.

@mih:

> It would be great to already have this implemented as an option that could take values, but doesn't have to. I have no good approach to suggest right now, though...

Moving to an option with values sounds fine to me.  My thinking was that using a bare option (until there are concrete ideas for other modes) doesn't block adding values because we could tack on optional values later (in the same way `status --annex` maps to `status --annex=basic`).

mih commented:
But it is difficult here. The command will pretty much always be last. I tried and the potential for confusion is high (swallow first item of a command).

kyleam commented:
> I tried and the potential for confusion is high (swallow first item of a command).

Ah, true, that's a good reason to avoid the optional value approach.

kyleam commented:
> It would be great to already have this implemented as an option that could take values, but doesn't have to

I've switched the option over to taking values and added a mode that shows only the expanded command.  Given the issue you pointed out with extended a flag later in this case, I think supporting values right away is the way to go, though now I'm on the fence about whether `--report` would be a better name than `--dry-run`.

kyleam commented:
The hirni failure is the same one that has been occurring in gh-5534.

https://github.com/datalad/datalad/pull/5539/checks?check_run_id=2254534141

<details>
<summary>error</summary>

```
datalad_hirni.tests.test_demos.test_demo_repro_analysis ... /tmp/datalad_temp_test_demo_repro_analysis9hlb8uij/sourcedata/code/hirni-toolbox/procedures/copy-converter.sh: line 9: $1: unbound variable
ERROR
[...]
======================================================================
ERROR: datalad_hirni.tests.test_demos.test_demo_repro_analysis
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/hostedtoolcache/Python/3.7.10/x64/lib/python3.7/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/opt/hostedtoolcache/Python/3.7.10/x64/lib/python3.7/site-packages/datalad/tests/utils.py", line 757, in _wrap_with_tempfile
    return t(*(arg + (filename,)), **kw)
  File "/opt/hostedtoolcache/Python/3.7.10/x64/lib/python3.7/site-packages/datalad/tests/utils.py", line 757, in _wrap_with_tempfile
    return t(*(arg + (filename,)), **kw)
  File "/opt/hostedtoolcache/Python/3.7.10/x64/lib/python3.7/site-packages/datalad_hirni/tests/utils.py", line 257, in newfunc
    return f(*(arg + (new_url,)), **kw)
  File "/opt/hostedtoolcache/Python/3.7.10/x64/lib/python3.7/site-packages/datalad_hirni/tests/test_demos.py", line 156, in test_demo_repro_analysis
    anonymize=True)
  File "/opt/hostedtoolcache/Python/3.7.10/x64/lib/python3.7/site-packages/datalad/distribution/dataset.py", line 505, in apply_func
    return f(**kwargs)
  File "/opt/hostedtoolcache/Python/3.7.10/x64/lib/python3.7/site-packages/datalad/interface/utils.py", line 482, in eval_func
    return return_func(generator_func)(*args, **kwargs)
  File "/opt/hostedtoolcache/Python/3.7.10/x64/lib/python3.7/site-packages/datalad/interface/utils.py", line 470, in return_func
    results = list(results)
  File "/opt/hostedtoolcache/Python/3.7.10/x64/lib/python3.7/site-packages/datalad/interface/utils.py", line 401, in generator_func
    allkwargs):
  File "/opt/hostedtoolcache/Python/3.7.10/x64/lib/python3.7/site-packages/datalad/interface/utils.py", line 557, in _process_results
    for res in results:
  File "/opt/hostedtoolcache/Python/3.7.10/x64/lib/python3.7/site-packages/datalad_hirni/commands/spec2bids.py", line 268, in __call__
    return_type='generator'
  File "/opt/hostedtoolcache/Python/3.7.10/x64/lib/python3.7/site-packages/datalad/interface/utils.py", line 401, in generator_func
    allkwargs):
  File "/opt/hostedtoolcache/Python/3.7.10/x64/lib/python3.7/site-packages/datalad/interface/utils.py", line 557, in _process_results
    for res in results:
  File "/opt/hostedtoolcache/Python/3.7.10/x64/lib/python3.7/site-packages/datalad/interface/run_procedure.py", line 451, in __call__
    return_type='generator'
  File "/opt/hostedtoolcache/Python/3.7.10/x64/lib/python3.7/site-packages/datalad/interface/utils.py", line 401, in generator_func
    allkwargs):
  File "/opt/hostedtoolcache/Python/3.7.10/x64/lib/python3.7/site-packages/datalad/interface/utils.py", line 557, in _process_results
    for res in results:
  File "/opt/hostedtoolcache/Python/3.7.10/x64/lib/python3.7/site-packages/datalad/core/local/run.py", line 275, in __call__
    dry_run=dry_run):
  File "/opt/hostedtoolcache/Python/3.7.10/x64/lib/python3.7/site-packages/datalad/core/local/run.py", line 791, in run_command
    raise exc
  File "/opt/hostedtoolcache/Python/3.7.10/x64/lib/python3.7/site-packages/datalad/core/local/run.py", line 556, in _execute_command
    command
  File "/opt/hostedtoolcache/Python/3.7.10/x64/lib/python3.7/site-packages/datalad/cmd.py", line 412, in run
    **results,
datalad.support.exceptions.CommandError: CommandError: 'bash /tmp/datalad_temp_test_demo_repro_analysis9hlb8uij/sourcedata/code/hirni-toolbox/procedures/copy-converter.sh ' failed with exitcode 1 under /tmp/datalad_temp_test_demo_repro_analysis9hlb8uij
```

</details>

yarikoptic commented:
windows fails seems to be legit but then why in only one of the runs?
```
======================================================================
FAIL: datalad.core.local.tests.test_run.test_dry_run
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\Python39-x64\lib\site-packages\nose\case.py", line 198, in runTest
    self.test(*self.arg)
  File "C:\Python39-x64\lib\site-packages\datalad\tests\utils.py", line 578, in _wrap_with_tree
    return t(*(arg + (d,)), **kw)
  File "C:\Python39-x64\lib\site-packages\datalad\core\local\tests\test_run.py", line 628, in test_dry_run
    assert_in(
AssertionError: 'blah "sub/baz"' not found in 'Dry run information\n location: C:\\DLTMP\\datalad_temp_tree_6m7w12oz\n\n expanded inputs:\n\n  [\'sub\\\\baz\']\n\n command:\n\n  blah {inputs}\n\n expanded command:\n\n  blah "sub\\baz"\n'
```

kyleam commented:
> windows fails seems to be legit

I somehow overlooked that failure.  Thanks!

> but then why in only one of the runs?

Of the Windows jobs, I think that's the only one that runs the `datalad.core` tests.

kyleam commented:
I've rebased to resolve conflicts and also appended "[dry-run]" to the action name to follow the recent change in `create-sibling-github`.

kyleam commented:
Thanks for the review, @yarikoptic.

@mih Are you okay with the current state, in particular with `--dry-run={basic,command}`?

yarikoptic commented:
eh, codespell already started to negatively effect the DX... hopefully would not happen that often ;-)

mih commented:
I like it! thx much!
